File: scripts/durable-fixtures.mjs
// SPDX-License-Identifier: Apache-2.0
import {readFileSync} from 'node:fs';
import {fixture as llmFixture} from './llm-fixtures.mjs';
import {DurableLlmLoop} from '@psp-cdl/llmproxy';
import {bindingDigest} from '@psp-cdl/mcpproxy';
import {StoreError} from '@psp-cdl/api-server/persistence';
export const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/llm/durable-0.1.json',import.meta.url),'utf8'));
export async function fixture(settings={}) {
  const f=await llmFixture({...settings,base:{...settings.base,durableTurns:!settings.disabledStore}}),events=[],commands=[];
  const authenticate=f.host.authenticate;
  f.host.authenticate=t=>{const p=authenticate(t);return p?{...p,scopes:[...p.scopes,...(f.flags.noReadScope?[]:['sessions:read']),...(f.flags.noWriteScope?[]:['sessions:write'])]}:null;};
  const host={...f.loopHost,
    planTurn:(_p,_b,d)=>{
      if(f.flags.throwPlan)throw new Error('PRIVATE_PLAN_DETAIL');
      if(f.flags.badPlan)return {state:{},complete:'yes',retained:{}};
      return {state:{answer:d.candidate.text,...(f.flags.forgedState?{llmCompletion:{policy:'lockdown'}}:{})},retained:{covenants:f.flags.retainedDisplay?['no-display-to-operator']:['no-training']},complete:!f.flags.continueTurn};
    },
    authorizeTransition:async(p,c)=>{
      commands.push(structuredClone(c.command));
      if(f.flags.transitionDrift)f.baseFlags.drift=true;
      if(f.flags.transitionCancel)f.baseFlags.cancelled=true;
      if(f.flags.transitionRevoke)f.baseFlags.revoked=true;
      if(f.flags.transitionExpiry)f.baseFlags.now=1700;
      if(f.flags.transitionRace){try{await f.update();}catch(e){if(e.code!=='STATE_BUSY')throw e;events.push({signal:'STATE_BUSY'});}}
      if(f.flags.mutateTransition){p.subjectId='forged';c.command.state={forged:true};c.result.output.text='forged';}
      if(f.flags.throwTransition)throw new Error('PRIVATE_TRANSITION_DETAIL');
      return !f.flags.denyTransition;
    },
    audit:(_p,event)=>{events.push(event);if(f.flags.throwAudit)throw new Error('PRIVATE_AUDIT_DETAIL');return !f.flags.denyAudit;},
    authorizeRecovery:()=>{if(f.flags.recoveryDrift)f.baseFlags.drift=true;if(f.flags.recoveryRevoke)f.baseFlags.revoked=true;return !f.flags.denyRecovery;},
    recoveryPolicy:(_p,b,r)=>({bindingDigest:f.flags.wrongRecoveryBinding?'wrong':bindingDigest(b),resources:f.flags.emptyRecovery?[]:[{classes:[],covenants:f.flags.unsupportedRecovery?['unknown-term']:r.retained.covenants,capabilities:f.flags.denyRecoveryPolicy?['used-for-model-training']:[],checks:{},parameters:{},context:{}}]}),
    policy:async(p,b,d,phase)=>{
      const result=await f.loopHost.policy(p,b,d,phase);
      if(f.flags.denyAfterCommit&&b.committedVersion)result.resources[0].capabilities=['used-for-model-training'];
      return result;
    }
  };
  const commit=f.backend.commit.bind(f.backend);
  f.backend.commit=async(...args)=>{
    const turn=args[2].some(w=>w.body.profile==='PSP-LLM-DURABLE-0.1');
    if(turn&&f.flags.failBeforeCommit)throw new StoreError('STORE_FAILURE');
    const ok=await commit(...args);
    if(turn&&ok){if(f.flags.cancelAfterCommit)f.baseFlags.cancelled=true;if(f.flags.driftAfterCommit)f.baseFlags.drift=true;if(f.flags.failAfterCommit)throw new StoreError('STORE_FAILURE');}
    return ok;
  };
  f.baseFlags.denyPersistence=!!f.flags.denyStorage;
  try {
    const loop=new DurableLlmLoop(f.store,f.gate,host,f.provider,{postCompletion:settings.postCompletion??'lockdown'});
    return {...f,loop,host,events,commands,options:{...f.options,requestId:'turn-1',expectedVersion:1}};
  }catch(e){f.close();throw e;}
}
export async function runCase(c) {
  let f;const steps=[];
  try {
    f=await fixture(c.settings);
    for(const step of c.steps??[{action:'run'}]) {
      Object.assign(f.flags,step.flags??{});Object.assign(f.baseFlags,step.baseFlags??{});
      const options={...f.options,...step.options},token=step.token??'test-owner';
      try {
        const value=step.action==='recover'?await f.loop.recover(token,f.session.sessionId,step.requestId??'turn-1',options):
          await f.loop.run(token,f.session.sessionId,step.request??{message:'Read synthetic data.'},options);
        steps.push({code:'OK',value});
      }catch(e){steps.push({code:e.code??'UNEXPECTED_ERROR',...(e.response?{response:e.response}:{})});}
    }
    const state=await f.store.execute(f.actor,{action:'getSession',sessionId:f.session.sessionId});
    return normalize({steps,codes:steps.map(s=>s.code),released:steps.filter(s=>s.code==='OK').length,...f.stats(),version:state.version,status:state.status,completion:state.llmCompletion??null,state:state.state,events:f.events,commands:f.commands,audits:f.events.filter(e=>e.signal==='post_completion_override_attempt').length,transitions:f.commands.length},f.session.sessionId);
  }catch(e){return {code:e.code??'UNEXPECTED_ERROR'};}
  finally{f?.close();}
}
export function normalize(value,sessionId){return JSON.parse(JSON.stringify(value).replaceAll(sessionId,'<session>'));}

File: scripts/durable_fixtures.py
# SPDX-License-Identifier: Apache-2.0
import json
from pathlib import Path
from copy import deepcopy
from llm_fixtures import Fixture as LlmFixture
from psp_cdl_llmproxy import DurableLlmLoop
from psp_cdl_mcpproxy import binding_digest
from psp_cdl_api_server.persistence import StoreError

SUITE=json.loads((Path(__file__).resolve().parents[1]/"conformance/vectors/llm/durable-0.1.json").read_text(encoding="utf-8"))


class Fixture(LlmFixture):
    def __init__(self, settings=None):
        settings=settings or {}
        super().__init__({**settings,"base":{**settings.get("base",{}),"durableTurns":not settings.get("disabledStore")}})
        self.events,self.commands=[],[]
        self.base.principal["scopes"] += ["sessions:read","sessions:write"]
        self.options.update(requestId="turn-1",expectedVersion=1)
        commit=self.base.backend.commit
        def committing(*args):
            turn=any(w["body"].get("profile")=="PSP-LLM-DURABLE-0.1" for w in args[2])
            if turn and self.flags.get("failBeforeCommit"): raise StoreError("STORE_FAILURE")
            ok=commit(*args)
            if turn and ok:
                if self.flags.get("cancelAfterCommit"): self.base.flags["cancelled"]=True
                if self.flags.get("driftAfterCommit"): self.base.flags["drift"]=True
                if self.flags.get("failAfterCommit"): raise StoreError("STORE_FAILURE")
            return ok
        self.base.backend.commit=committing
        self.base.flags["denyPersistence"]=bool(self.flags.get("denyStorage"))
        try: self.loop=DurableLlmLoop(self.base.store,self.base.gate,self,self.provider,{"postCompletion":settings.get("postCompletion","lockdown")})
        except Exception:
            self.close()
            raise
    def plan_turn(self,p,b,data):
        if self.flags.get("throwPlan"): raise RuntimeError("PRIVATE_PLAN_DETAIL")
        if self.flags.get("badPlan"): return {"state":{},"complete":"yes","retained":{}}
        return {"state":{"answer":data["candidate"]["text"],**({"llmCompletion":{"policy":"lockdown"}} if self.flags.get("forgedState") else {})},"retained":{"covenants":["no-display-to-operator"] if self.flags.get("retainedDisplay") else ["no-training"]},"complete":not self.flags.get("continueTurn")}
    def authenticate(self,token):
        principal=super().authenticate(token)
        if principal:
            principal=deepcopy(principal)
            principal["scopes"]=[s for s in principal["scopes"] if not (s=="sessions:read" and self.flags.get("noReadScope") or s=="sessions:write" and self.flags.get("noWriteScope") or s=="models:invoke" and self.flags.get("noModelScope"))]
        return principal
    def authorize_transition(self,p,c):
        self.commands.append(deepcopy(c["command"]))
        for flag,key,value in (("transitionDrift","drift",True),("transitionCancel","cancelled",True),("transitionRevoke","revoked",True),("transitionExpiry","now",1700)):
            if self.flags.get(flag): self.base.flags[key]=value
        if self.flags.get("transitionRace"):
            try: self.base.update()
            except StoreError as exc:
                if exc.code!="STATE_BUSY": raise
                self.events.append({"signal":"STATE_BUSY"})
        if self.flags.get("mutateTransition"):
            p["subjectId"],c["command"]["state"],c["result"]["output"]["text"]="forged",{"forged":True},"forged"
        if self.flags.get("throwTransition"): raise RuntimeError("PRIVATE_TRANSITION_DETAIL")
        return not self.flags.get("denyTransition")
    def audit(self,p,event):
        self.events.append(event)
        if self.flags.get("throwAudit"): raise RuntimeError("PRIVATE_AUDIT_DETAIL")
        return not self.flags.get("denyAudit")
    def authorize_recovery(self,*_):
        if self.flags.get("recoveryDrift"): self.base.flags["drift"]=True
        if self.flags.get("recoveryRevoke"): self.base.flags["revoked"]=True
        return not self.flags.get("denyRecovery")
    def recovery_policy(self,p,b,r):
        return {"bindingDigest":"wrong" if self.flags.get("wrongRecoveryBinding") else binding_digest(b),"resources":[] if self.flags.get("emptyRecovery") else [{"classes":[],"covenants":["unknown-term"] if self.flags.get("unsupportedRecovery") else r["retained"]["covenants"],"capabilities":["used-for-model-training"] if self.flags.get("denyRecoveryPolicy") else [],"checks":{},"parameters":{},"context":{}}]}
    def policy(self,p,b,d,phase):
        result=super().policy(p,b,d,phase)
        if self.flags.get("denyAfterCommit") and b.get("committedVersion"): result["resources"][0]["capabilities"]=["used-for-model-training"]
        return result


def run_case(case):
    f=None
    steps=[]
    try:
        f=Fixture(case.get("settings"))
        for step in case.get("steps",[{"action":"run"}]):
            f.flags.update(step.get("flags",{})); f.base.flags.update(step.get("baseFlags",{}))
            options={**f.options,**step.get("options",{})}
            token=step.get("token","test-owner")
            try:
                value=f.loop.recover(token,f.base.session["sessionId"],step.get("requestId","turn-1"),options) if step["action"]=="recover" else f.loop.run(token,f.base.session["sessionId"],step.get("request",{"message":"Read synthetic data."}),options)
                steps.append({"code":"OK","value":value})
            except Exception as exc:
                steps.append({"code":getattr(exc,"code","UNEXPECTED_ERROR"),**({"response":exc.response} if hasattr(exc,"response") else {})})
        state=f.base.store.execute(f.base.actor,{"action":"getSession","sessionId":f.base.session["sessionId"]})
        return normalize({"steps":steps,"codes":[s["code"] for s in steps],"released":sum(s["code"]=="OK" for s in steps),**f.stats(),"version":state["version"],"status":state["status"],"completion":state.get("llmCompletion"),"state":state["state"],"events":f.events,"commands":f.commands,"audits":sum(e["signal"]=="post_completion_override_attempt" for e in f.events),"transitions":len(f.commands)},f.base.session["sessionId"])
    except Exception as exc: return {"code":getattr(exc,"code","UNEXPECTED_ERROR")}
    finally:
        if f: f.close()


def normalize(value,session_id): return json.loads(json.dumps(value).replace(session_id,"<session>"))
